File: src/main/java/com/milaboratory/mixcr/trees/TreeBuilderByAncestors.java
package com.milaboratory.mixcr.trees;

import java.math.BigDecimal;
import java.util.Comparator;
import java.util.List;
import java.util.Objects;
import java.util.Optional;
import java.util.function.BiFunction;
import java.util.function.Function;
import java.util.stream.Collectors;
import java.util.stream.Stream;

import static java.util.Comparator.comparing;

public class TreeBuilderByAncestors<T, E, M> {
    public static final Comparator<Action> ACTION_COMPARATOR = Comparator
            .<Action, BigDecimal>comparing(action -> action.changeOfDistance().add(action.distanceFromObserved()))
            .thenComparing(Action::changeOfDistance)
            .thenComparing(action -> action instanceof TreeBuilderByAncestors.Insert ? 1 : -1);

    private final BiFunction<E, M, BigDecimal> distance;
    /**
     * parent, observed -> reconstructed
     */
    private final Function<T, E> asAncestor;
    private final BiFunction<E, E, M> mutationsBetween;
    private final BiFunction<E, M, E> mutate;
    private final BiFunction<M, M, M> findCommonMutations;
    private final BiFunction<E, E, E> postprocessDescendants;

    private final Tree<ObservedOrReconstructed<T, E>> tree;
    private final int countOfNodesToProbe;
    private int counter;

    public TreeBuilderByAncestors(
            E root,
            BiFunction<E, M, BigDecimal> distance,
            BiFunction<E, E, M> mutationsBetween,
            BiFunction<E, M, E> mutate,
            Function<T, E> asAncestor,
            BiFunction<M, M, M> findCommonMutations,
            BiFunction<E, E, E> postprocessDescendants,
            int countOfNodesToProbe
    ) {
        this(
                distance,
                asAncestor,
                mutationsBetween,
                mutate,
                findCommonMutations,
                postprocessDescendants,
                new Tree<>(new Tree.Node<>(new Reconstructed<>(root, BigDecimal.ZERO, 0))),
                countOfNodesToProbe,
                1
        );
    }

    private TreeBuilderByAncestors(
            BiFunction<E, M, BigDecimal> distance,
            Function<T, E> asAncestor,
            BiFunction<E, E, M> mutationsBetween,
            BiFunction<E, M, E> mutate,
            BiFunction<M, M, M> findCommonMutations,
            BiFunction<E, E, E> postprocessDescendants,
            Tree<ObservedOrReconstructed<T, E>> tree,
            int countOfNodesToProbe,
            int counter
    ) {
        this.distance = distance;
        this.asAncestor = asAncestor;
        this.mutationsBetween = mutationsBetween;
        this.mutate = mutate;
        this.findCommonMutations = findCommonMutations;
        this.postprocessDescendants = postprocessDescendants;
        this.tree = tree;
        this.countOfNodesToProbe = countOfNodesToProbe;
        this.counter = counter;
    }

    public TreeBuilderByAncestors<T, E, M> copy() {
        return new TreeBuilderByAncestors<>(
                distance,
                asAncestor,
                mutationsBetween,
                mutate,
                findCommonMutations,
                postprocessDescendants,
                tree.copy(),
                countOfNodesToProbe,
                counter
        );
    }

    public TreeBuilderByAncestors<T, E, M> addNode(T toAdd) {
        Action bestAction = bestActionForObserved(toAdd);
        bestAction.apply();
        return this;
    }

    public Action bestActionForObserved(T toAdd) {
        E addedAsAncestor = asAncestor.apply(toAdd);
        return bestAction(addedAsAncestor, distanceFromParent -> {
            Tree.Node<ObservedOrReconstructed<T, E>> nodeToAdd;
            if (distanceFromParent.compareTo(BigDecimal.ZERO) == 0) {
                nodeToAdd = new Tree.Node<>(new Observed<>(toAdd, ++counter));
            } else {
                nodeToAdd = new Tree.Node<>(new Reconstructed<>(addedAsAncestor, BigDecimal.ZERO, ++counter));
                nodeToAdd.addChild(new Tree.Node<>(new Observed<>(toAdd, ++counter)), BigDecimal.ZERO);
            }
            return nodeToAdd;
        });
    }

    /**
     * Assumptions:
     * 1. Nodes must be added in order by distance from the root
     * 2. findCommonAncestor is constructing ancestor based on root (left common mutations on nodes from root)
     * <p>
     * <p>
     * Invariants of the tree:
     * 1. Root node is always reconstructed. Root is not changing
     * 2. Leaves are always observed.
     * 3. Distance between observed and reconstructed could be zero (they are identical).
     * 4. Distance between observed can't be zero.
     * 5. Siblings have no common ancestors.
     */
    private Action bestAction(E addedAsAncestor, Function<BigDecimal, Tree.Node<ObservedOrReconstructed<T, E>>> nodeGenerator) {
        List<Tree.NodeWithParent<ObservedOrReconstructed<T, E>>> nearestNodes = tree.allNodes()
                .filter(it -> it.getNode().getContent() instanceof Reconstructed<?, ?>)
                .sorted(Comparator.comparing(compareWith -> {
                    Reconstructed<T, E> reconstructed = (Reconstructed<T, E>) compareWith.getNode().getContent();
                    E nodeContent = reconstructed.getContent();
                    return distance.apply(
                            nodeContent,
                            mutationsBetween.apply(nodeContent, addedAsAncestor)
                    ).add(reconstructed.getMinDistanceFromObserved());
                }))
                .limit(countOfNodesToProbe)
                .collect(Collectors.toList());
        Stream<Action> possibleActions = nearestNodes.stream().flatMap(nodeWithParent -> {
            Tree.Node<ObservedOrReconstructed<T, E>> chosenNode = nodeWithParent.getNode();
            //search for siblings with common mutations with the added node
            Optional<Replace> siblingToMergeWith = chosenNode.getLinks().stream()
                    .filter(it -> it.getNode().getContent() instanceof Reconstructed<?, ?>)
                    .map(link -> replaceChild(chosenNode, link.getNode(), link.getDistance(), addedAsAncestor, nodeGenerator))
                    .filter(Objects::nonNull)
                    //choose a sibling with max score of common mutations
                    .max(comparing(Replace::getDistanceFromParentToCommon));
            if (siblingToMergeWith.isPresent()) {
                //the added node (A), the chosen node (B), the sibling (D) has common ancestor (C) with A
                //
                //       P                   P
                //       |                   |
                //     --*--               --*--
                //     |   |               |   |
                //     K   B      ==>      K   B
                //         |                   |
                //       --*--               --*--
                //       |   |               |   |
                //       R   D               R   C
                //                               |
                //                             --*--
                //                             |   |
                //                             A   D
                //
                // R and C has no common ancestors because R and D haven't one
                return siblingToMergeWith.stream();
            } else {
                //the added node (A), the chosen node (B), no siblings with common ancestors
                //
                //       P                   P
                //       |                   |
                //     --*--               --*--
                //     |   |               |   |
                //     K   B      ==>      K   B
                //         |                   |
                //       --*--               --*-----
                //       |   |               |   |  |
                //       R   T               R   T  A
                Stream<Replace> insertAsParent;
                if (nodeWithParent.getParent() != null) {
                    insertAsParent = Stream.of(replaceChild(
                            nodeWithParent.getParent(),
                            chosenNode,
                            nodeWithParent.getDistance(),
                            addedAsAncestor,
                            nodeGenerator
                    )).filter(Objects::nonNull);
                } else {
                    insertAsParent = Stream.empty();
                }
                return Stream.concat(
                        Stream.of(insertAsDirectDescendant(chosenNode, addedAsAncestor, nodeGenerator)),
                        insertAsParent
                );
            }
        });
        //optimize sum of distances from observed nodes
        List<Action> temp = possibleActions.collect(Collectors.toList());
        return temp.stream().min(ACTION_COMPARATOR).orElseThrow(IllegalArgumentException::new);
    }

    public BigDecimal distanceFromRootToObserved(T node) {
        E rootContent = ((Reconstructed<T, E>) tree.getRoot().getContent()).getContent();
        return distance.apply(
                rootContent,
                mutationsBetween.apply(rootContent, asAncestor.apply(node))
        );
    }

    private Insert insertAsDirectDescendant(Tree.Node<ObservedOrReconstructed<T, E>> parent, E toAdd, Function<BigDecimal, Tree.Node<ObservedOrReconstructed<T, E>>> nodeGenerator) {
        E contentOfParent = ((Reconstructed<T, E>) parent.getContent()).getContent();
        BigDecimal distanceBetweenParentAndAdded = distance.apply(contentOfParent, mutationsBetween.apply(contentOfParent, toAdd));
        Tree.Node<ObservedOrReconstructed<T, E>> nodeToInsert = nodeGenerator.apply(distanceBetweenParentAndAdded);
        return new Insert(
                parent,
                distanceBetweenParentAndAdded,
                nodeToInsert
        );
    }

    /**
     * Replacing node with subtree with common ancestor and leaves as added node and node that was here before
     *
     * <pre>
     *              P                   P
     *              |                   |
     *            --*--               --*--
     *            |   |    ==>        |   |
     *            R   B               R   C
     *                                    |
     *                                  --*--
     *                                  |   |
     *                                  A   B
     * </pre>
     * P - parent
     * A - toAdd
     * B - child
     * C - commonAncestor
     */
    private Replace replaceChild(
            Tree.Node<ObservedOrReconstructed<T, E>> parent,
            Tree.Node<ObservedOrReconstructed<T, E>> child,
            BigDecimal distanceFromParentToChild,
            E addedAsAncestor,
            Function<BigDecimal, Tree.Node<ObservedOrReconstructed<T, E>>> nodeGenerator
    ) {
        Reconstructed<T, E> parentAsReconstructed = (Reconstructed<T, E>) parent.getContent();
        E parentContent = parentAsReconstructed.getContent();
        E childContent = ((Reconstructed<T, E>) child.getContent()).getContent();

        M mutationsToAdded = mutationsBetween.apply(parentContent, addedAsAncestor);
        M mutationsToChild = mutationsBetween.apply(parentContent, childContent);

        M commonMutations = findCommonMutations.apply(mutationsToAdded, mutationsToChild);
        BigDecimal distanceFromParentToCommon = distance.apply(parentContent, commonMutations);
        //if distance is zero than there is no common mutations
        if (distanceFromParentToCommon.compareTo(BigDecimal.ZERO) == 0) {
            return null;
        }

        E commonAncestor = mutate.apply(parentContent, commonMutations);
        M fromCommonToChild = mutationsBetween.apply(commonAncestor, childContent);
        BigDecimal distanceFromCommonAncestorToChild = distance.apply(commonAncestor, fromCommonToChild);
        //if distance is zero than result of replacement equals to insertion
        if (distanceFromCommonAncestorToChild.compareTo(BigDecimal.ZERO) == 0) {
            return null;
        }
        BigDecimal distanceFromCommonToAdded = distance.apply(commonAncestor, mutationsBetween.apply(commonAncestor, addedAsAncestor));

        BigDecimal minDistanceFromObserved = Stream.of(
                distanceFromCommonToAdded,
                distanceFromCommonAncestorToChild,
                parentAsReconstructed.getMinDistanceFromObserved().add(distanceFromParentToCommon)
        ).min(BigDecimal::compareTo).get();
        Tree.Node<ObservedOrReconstructed<T, E>> replacement = new Tree.Node<>(new Reconstructed<>(
                commonAncestor,
                minDistanceFromObserved,
                ++counter
        ));
        Tree.Node<ObservedOrReconstructed<T, E>> rebuiltChild = new Tree.Node<>(new Reconstructed<>(
                mutate.apply(commonAncestor, fromCommonToChild),
                ((Reconstructed<T, E>) child.getContent()).getMinDistanceFromObserved(),
                ++counter
        ));
        child.getLinks().forEach(link -> rebuiltChild.addChild(link.getNode(), link.getDistance()));
        replacement.addChild(rebuiltChild, distanceFromCommonAncestorToChild);

        replacement.addChild(nodeGenerator.apply(distanceFromCommonToAdded), distanceFromCommonToAdded);

        return new Replace(
                parent,
                child,
                replacement,
                distanceFromParentToCommon,
                distanceFromCommonAncestorToChild.add(distanceFromCommonToAdded),
                distanceFromParentToChild
        );
    }

    public Tree<ObservedOrReconstructed<T, E>> getTree() {
        return tree;
    }

    public static abstract class Action {
        protected abstract BigDecimal changeOfDistance();

        protected abstract BigDecimal distanceFromObserved();

        protected abstract void apply();
    }

    private class Insert extends Action {
        private final Tree.Node<ObservedOrReconstructed<T, E>> parent;
        private final BigDecimal distanceFromParentAndInsertion;
        private final Tree.Node<ObservedOrReconstructed<T, E>> nodeToInsert;

        public Insert(Tree.Node<ObservedOrReconstructed<T, E>> parent, BigDecimal distanceFromParentAndInsertion, Tree.Node<ObservedOrReconstructed<T, E>> nodeToInsert) {
            this.parent = parent;
            this.distanceFromParentAndInsertion = distanceFromParentAndInsertion;
            this.nodeToInsert = nodeToInsert;
        }

        @Override
        protected BigDecimal changeOfDistance() {
            return distanceFromParentAndInsertion;
        }

        @Override
        protected BigDecimal distanceFromObserved() {
            return ((Reconstructed<T, E>) parent.getContent()).getMinDistanceFromObserved();
        }

        @Override
        protected void apply() {
            Reconstructed<T, E> parentAsReconstructed = (Reconstructed<T, E>) parent.getContent();
            if (distanceFromParentAndInsertion.compareTo(BigDecimal.ZERO) == 0) {
                parentAsReconstructed.setMinDistanceFromObserved(BigDecimal.ZERO);
            }
            parent.addChild(nodeToInsert, distanceFromParentAndInsertion);
        }
    }

    private class Replace extends Action {
        private final Tree.Node<ObservedOrReconstructed<T, E>> parent;
        private final Tree.Node<ObservedOrReconstructed<T, E>> replaceWhat;
        private final Tree.Node<ObservedOrReconstructed<T, E>> replacement;
        private final BigDecimal distanceFromParentToCommon;
        private final BigDecimal sumOfDistancesFromAncestor;
        private final BigDecimal distanceFromParentToReplaceWhat;

        public Replace(Tree.Node<ObservedOrReconstructed<T, E>> parent,
                       Tree.Node<ObservedOrReconstructed<T, E>> replaceWhat,
                       Tree.Node<ObservedOrReconstructed<T, E>> replacement,
                       BigDecimal distanceFromParentToCommon,
                       BigDecimal sumOfDistancesFromAncestor,
                       BigDecimal distanceFromParentToReplaceWhat
        ) {
            this.parent = parent;
            this.replaceWhat = replaceWhat;
            this.replacement = replacement;
            this.distanceFromParentToCommon = distanceFromParentToCommon;
            this.sumOfDistancesFromAncestor = sumOfDistancesFromAncestor;
            this.distanceFromParentToReplaceWhat = distanceFromParentToReplaceWhat;
        }

        BigDecimal getDistanceFromParentToCommon() {
            return distanceFromParentToCommon;
        }

        @Override
        protected BigDecimal changeOfDistance() {
            return distanceFromParentToCommon.subtract(distanceFromParentToReplaceWhat).add(sumOfDistancesFromAncestor);
        }

        @Override
        protected BigDecimal distanceFromObserved() {
            return ((Reconstructed<T, E>) parent.getContent()).getMinDistanceFromObserved();
        }

        @Override
        protected void apply() {
            parent.replaceChild(
                    replaceWhat,
                    postprocess(replacement),
                    distanceFromParentToCommon
            );
        }

        private Tree.Node<ObservedOrReconstructed<T, E>> postprocess(Tree.Node<ObservedOrReconstructed<T, E>> parentToPostprocess) {
            return new Tree.Node<>(
                    parentToPostprocess.getContent(),
                    postprocess(
                            ((Reconstructed<T, E>) parentToPostprocess.getContent()).getContent(),
                            parentToPostprocess.getLinks()
                    )
            );
        }

        private List<Tree.NodeLink<ObservedOrReconstructed<T, E>>> postprocess(
                E parentContent, List<Tree.NodeLink<ObservedOrReconstructed<T, E>>> links
        ) {
            return links.stream()
                    .map(link -> {
                        Tree.Node<ObservedOrReconstructed<T, E>> child = link.getNode();
                        if (child.getContent() instanceof Reconstructed<?, ?>) {
                            Reconstructed<T, E> childContent = (Reconstructed<T, E>) child.getContent();
                            E mapped = postprocessDescendants.apply(parentContent, childContent.getContent());
                            return new Tree.NodeLink<>(
                                    new Tree.Node<>(
                                            new Reconstructed<>(
                                                    mapped,
                                                    //TODO recalculate minDistanceFromObserved
                                                    childContent.getMinDistanceFromObserved(),
                                                    ++counter
                                            ),
                                            postprocess(mapped, child.getLinks())
                                    ),
                                    distance.apply(parentContent, mutationsBetween.apply(parentContent, mapped))
                            );
                        } else {
                            return link;
                        }
                    })
                    .collect(Collectors.toList());
        }
    }

    public static abstract class ObservedOrReconstructed<T, E> {
        private final int id;

        protected ObservedOrReconstructed(int id) {
            this.id = id;
        }

        public int getId() {
            return id;
        }

        public <R1, R2> ObservedOrReconstructed<R1, R2> map(Function<T, R1> mapObserved, Function<E, R2> mapReconstructed) {
            if (this instanceof Observed<?, ?>) {
                return new Observed<>(mapObserved.apply(((Observed<T, E>) this).getContent()), id);
            } else if (this instanceof Reconstructed<?, ?>) {
                Reconstructed<T, E> casted = (Reconstructed<T, E>) this;
                return new Reconstructed<>(mapReconstructed.apply(casted.getContent()), casted.getMinDistanceFromObserved(), id);
            } else {
                throw new IllegalArgumentException();
            }
        }

        public <R> R convert(Function<T, R> mapObserved, Function<E, R> mapReconstructed) {
            if (this instanceof Observed<?, ?>) {
                return mapObserved.apply(((Observed<T, E>) this).getContent());
            } else if (this instanceof Reconstructed<?, ?>) {
                return mapReconstructed.apply(((Reconstructed<T, E>) this).getContent());
            } else {
                throw new IllegalArgumentException();
            }
        }
    }

    static class Observed<T, E> extends ObservedOrReconstructed<T, E> {
        private final T content;

        private Observed(T content, int id) {
            super(id);
            this.content = content;
        }

        public T getContent() {
            return content;
        }
    }

    static class Reconstructed<T, E> extends ObservedOrReconstructed<T, E> {
        private final E content;
        /**
         * This number represents distance that will be added after removing of all reconstructed nodes from the tree.
         * <p>
         * zero for root
         * zero if had an observed child with distance zero
         * otherwise it is a distance from the nearest parent with minDistanceFromObserved equals zero
         */
        private BigDecimal minDistanceFromObserved;

        private Reconstructed(E content, BigDecimal minDistanceFromObserved, int id) {
            super(id);
            this.content = content;
            this.minDistanceFromObserved = minDistanceFromObserved;
        }

        private void setMinDistanceFromObserved(BigDecimal minDistanceFromObserved) {
            this.minDistanceFromObserved = minDistanceFromObserved;
        }

        BigDecimal getMinDistanceFromObserved() {
            return minDistanceFromObserved;
        }

        public E getContent() {
            return content;
        }
    }
}

File: src/main/kotlin/com/milaboratory/mixcr/trees/TreeBuilderByAncestors.kt
/*
 * Copyright (c) 2014-2022, MiLaboratories Inc. All Rights Reserved
 *
 * Before downloading or accessing the software, please read carefully the
 * License Agreement available at:
 * https://github.com/milaboratory/mixcr/blob/develop/LICENSE
 *
 * By downloading or accessing the software, you accept and agree to be bound
 * by the terms of the License Agreement. If you do not want to agree to the terms
 * of the Licensing Agreement, you must not download or access the software.
 */
package com.milaboratory.mixcr.trees

import com.milaboratory.mixcr.trees.Tree.NodeLink

//TODO why abstraction
class TreeBuilderByAncestors<T, E, M> private constructor(
    private val distance: (E, M) -> Double,
    private val asAncestor: (T) -> E,
    private val mutationsBetween: (E, E) -> M,
    private val mutate: (E, M) -> E,
    private val findCommonMutations: (M, M) -> M,
    private val postprocessDescendants: (parent: E, child: E) -> E,
    val tree: Tree<ObservedOrReconstructed<T, E>>,
    private val countOfNodesToProbe: Int,
    private var counter: Int
) {
    constructor(
        root: E,
        distance: (E, M) -> Double,
        mutationsBetween: (E, E) -> M,
        mutate: (E, M) -> E,
        asAncestor: (T) -> E,
        findCommonMutations: (M, M) -> M,
        postprocessDescendants: (E, E) -> E,
        countOfNodesToProbe: Int
    ) : this(
        distance,
        asAncestor,
        mutationsBetween,
        mutate,
        findCommonMutations,
        postprocessDescendants,
        Tree(Tree.Node(Reconstructed(root, 0.0, 0))),
        countOfNodesToProbe,
        1
    )

    fun copy(): TreeBuilderByAncestors<T, E, M> = TreeBuilderByAncestors(
        distance,
        asAncestor,
        mutationsBetween,
        mutate,
        findCommonMutations,
        postprocessDescendants,
        tree.copy(),
        countOfNodesToProbe,
        counter
    )

    fun addNode(toAdd: T): TreeBuilderByAncestors<T, E, M> {
        val bestAction = bestActionForObserved(toAdd)
        bestAction.apply()
        return this
    }

    fun bestActionForObserved(toAdd: T): Action<E> {
        val addedAsAncestor = asAncestor(toAdd)
        return bestAction(addedAsAncestor) { distanceFromParent ->
            val nodeToAdd: Tree.Node<ObservedOrReconstructed<T, E>>
            if (distanceFromParent == 0.0) {
                nodeToAdd = Tree.Node(Observed(toAdd, ++counter))
            } else {
                nodeToAdd = Tree.Node(Reconstructed(addedAsAncestor, 0.0, ++counter))
                nodeToAdd.addChild(Tree.Node(Observed(toAdd, ++counter)), 0.0)
            }
            nodeToAdd
        }
    }

    /**
     * Assumptions:
     * 1. Nodes must be added in order by distance from the root
     * 2. findCommonAncestor is constructing ancestor based on root (left common mutations on nodes from root)
     *
     *
     *
     *
     * Invariants of the tree:
     * 1. Root node is always reconstructed. Root is not changing
     * 2. Leaves are always observed.
     * 3. Distance between observed and reconstructed could be zero (they are identical).
     * 4. Distance between observed can't be zero.
     * 5. Siblings have no common ancestors.
     */
    private fun bestAction(
        addedAsAncestor: E,
        //TODO inline
        nodeGenerator: (Double) -> Tree.Node<ObservedOrReconstructed<T, E>>
    ): Action<E> {
        val nearestNodes = tree.allNodes()
            .asSequence()
            .filter { it.node.content is Reconstructed<*, *> }
            .map { compareWith ->
                val reconstructed = compareWith.node.content as Reconstructed<T, E>
                val nodeContent = reconstructed.content
                compareWith to distance(
                    nodeContent,
                    mutationsBetween(nodeContent, addedAsAncestor)
                ) + reconstructed.minDistanceFromObserved
            }
            .sortedBy { it.second }
            .map { it.first }
            .take(countOfNodesToProbe)
            .toList()
        val possibleActions = nearestNodes.flatMap { nodeWithParent ->
            val chosenNode = nodeWithParent.node
            //search for siblings with common mutations with the added node
            val siblingToMergeWith: Replace? = chosenNode.links
                .filter { it.node.content is Reconstructed<*, *> }
                //choose a sibling with max score of common mutations
                .mapNotNull { link ->
                    replaceChild(
                        chosenNode,
                        link.node,
                        link.distance,
                        addedAsAncestor,
                        nodeGenerator
                    )
                }
                .maxByOrNull { it.distanceFromParentToCommon }
            if (siblingToMergeWith != null) {
                //the added node (A), the chosen node (B), the sibling (D) has common ancestor (C) with A
                //
                //       P                   P
                //       |                   |
                //     --*--               --*--
                //     |   |               |   |
                //     K   B      ==>      K   B
                //         |                   |
                //       --*--               --*--
                //       |   |               |   |
                //       R   D               R   C
                //                               |
                //                             --*--
                //                             |   |
                //                             A   D
                //
                // R and C has no common ancestors because R and D haven't one
                return@flatMap listOf(siblingToMergeWith)
            } else {
                //the added node (A), the chosen node (B), no siblings with common ancestors
                //
                //       P                   P
                //       |                   |
                //     --*--               --*--
                //     |   |               |   |
                //     K   B      ==>      K   B
                //         |                   |
                //       --*--               --*-----
                //       |   |               |   |  |
                //       R   T               R   T  A
                val insertAsParent = if (nodeWithParent.parent != null) {
                    replaceChild(
                        nodeWithParent.parent,
                        chosenNode,
                        nodeWithParent.distance!!,
                        addedAsAncestor,
                        nodeGenerator
                    )
                } else {
                    null
                }
                return@flatMap listOfNotNull(
                    insertAsDirectDescendant(chosenNode, addedAsAncestor, nodeGenerator),
                    insertAsParent
                )
            }
        }
        //optimize sum of distances from observed nodes
        return possibleActions.minWithOrNull(ACTION_COMPARATOR)!!
    }

    fun distanceFromRootToObserved(node: T): Double {
        val rootContent = (tree.root.content as Reconstructed<T, E>).content
        return distance(
            rootContent,
            mutationsBetween(rootContent, asAncestor(node))
        )
    }

    private fun insertAsDirectDescendant(
        parent: Tree.Node<ObservedOrReconstructed<T, E>>,
        toAdd: E,
        nodeGenerator: (Double) -> Tree.Node<ObservedOrReconstructed<T, E>>
    ): Insert {
        val contentOfParent = (parent.content as Reconstructed<T, E>).content
        val distanceBetweenParentAndAdded =
            distance(contentOfParent, mutationsBetween(contentOfParent, toAdd))
        val nodeToInsert = nodeGenerator(distanceBetweenParentAndAdded)
        return Insert(
            parent,
            distanceBetweenParentAndAdded,
            nodeToInsert
        )
    }

    /**
     * Replacing node with subtree with common ancestor and leaves as added node and node that was here before
     *
     * <pre>
     * P                   P
     * |                   |
     * --*--               --*--
     * |   |    ==>        |   |
     * R   B               R   C
     * |
     * --*--
     * |   |
     * A   B
    </pre> *
     * P - parent
     * A - toAdd
     * B - child
     * C - commonAncestor
     */
    private fun replaceChild(
        parent: Tree.Node<ObservedOrReconstructed<T, E>>,
        child: Tree.Node<ObservedOrReconstructed<T, E>>,
        distanceFromParentToChild: Double,
        addedAsAncestor: E,
        nodeGenerator: (Double) -> Tree.Node<ObservedOrReconstructed<T, E>>
    ): Replace? {
        val parentAsReconstructed = parent.content as Reconstructed<T, E>
        val parentContent = parentAsReconstructed.content
        val childContent = (child.content as Reconstructed<T, E>).content
        val mutationsToAdded = mutationsBetween(parentContent, addedAsAncestor)
        val mutationsToChild = mutationsBetween(parentContent, childContent)
        val commonMutations = findCommonMutations(mutationsToAdded, mutationsToChild)
        val distanceFromParentToCommon = distance(parentContent, commonMutations)
        //if distance is zero than there is no common mutations
        if (distanceFromParentToCommon == 0.0) {
            return null
        }
        val commonAncestor = mutate(parentContent, commonMutations)
        val fromCommonToChild = mutationsBetween(commonAncestor, childContent)
        val distanceFromCommonAncestorToChild = distance(commonAncestor, fromCommonToChild)
        //if distance is zero than result of replacement equals to insertion
        if (distanceFromCommonAncestorToChild == 0.0) {
            return null
        }
        val distanceFromCommonToAdded = distance(commonAncestor, mutationsBetween(commonAncestor, addedAsAncestor))
        val minDistanceFromObserved = minOf(
            distanceFromCommonToAdded,
            //TODO check
            distanceFromCommonAncestorToChild,
            parentAsReconstructed.minDistanceFromObserved + distanceFromParentToCommon
        )
        val replacement = Tree.Node<ObservedOrReconstructed<T, E>>(
            Reconstructed(
                commonAncestor,
                minDistanceFromObserved,
                ++counter
            )
        )
        val rebuiltChild = Tree.Node<ObservedOrReconstructed<T, E>>(
            Reconstructed(
                mutate(commonAncestor, fromCommonToChild),
                child.content.minDistanceFromObserved,
                ++counter
            )
        )
        child.links.forEach { link ->
            rebuiltChild.addChild(
                link.node,
                link.distance
            )
        }
        replacement.addChild(rebuiltChild, distanceFromCommonAncestorToChild)
        replacement.addChild(nodeGenerator(distanceFromCommonToAdded), distanceFromCommonToAdded)
        return Replace(
            parent,
            child,
            replacement,
            distanceFromParentToCommon,
            distanceFromCommonAncestorToChild + distanceFromCommonToAdded,
            distanceFromParentToChild
        )
    }

    abstract class Action<E> {
        abstract val changeOfDistance: Double
        abstract val distanceFromObserved: Double
        abstract fun apply()

        abstract fun parentContent(): E
    }

    private inner class Insert(
        private val parent: Tree.Node<ObservedOrReconstructed<T, E>>,
        private val distanceFromParentAndInsertion: Double,
        private val nodeToInsert: Tree.Node<ObservedOrReconstructed<T, E>>
    ) : Action<E>() {
        override val changeOfDistance get() = distanceFromParentAndInsertion

        override val distanceFromObserved get() = (parent.content as Reconstructed<T, E>).minDistanceFromObserved

        override fun apply() {
            val parentAsReconstructed = parent.content as Reconstructed<T, E>
            if (distanceFromParentAndInsertion == 0.0) {
                parentAsReconstructed.minDistanceFromObserved = 0.0
            }
            parent.addChild(nodeToInsert, distanceFromParentAndInsertion)
        }

        override fun parentContent(): E = (parent.content as Reconstructed<T, E>).content
    }

    private inner class Replace(
        private val parent: Tree.Node<ObservedOrReconstructed<T, E>>,
        private val replaceWhat: Tree.Node<ObservedOrReconstructed<T, E>>,
        private val replacement: Tree.Node<ObservedOrReconstructed<T, E>>,
        val distanceFromParentToCommon: Double,
        private val sumOfDistancesFromAncestor: Double,
        private val distanceFromParentToReplaceWhat: Double
    ) : Action<E>() {
        override val changeOfDistance get() = distanceFromParentToCommon - distanceFromParentToReplaceWhat + sumOfDistancesFromAncestor

        override val distanceFromObserved get() = (parent.content as Reconstructed<T, E>).minDistanceFromObserved

        override fun apply() {
            parent.replaceChild(
                replaceWhat,
                postprocess(replacement),
                distanceFromParentToCommon
            )
        }

        override fun parentContent(): E = (parent.content as Reconstructed<T, E>).content

        private fun postprocess(parentToPostprocess: Tree.Node<ObservedOrReconstructed<T, E>>): Tree.Node<ObservedOrReconstructed<T, E>> {
            val result = Tree.Node(
                parentToPostprocess.content
            )
            postprocess(
                (parentToPostprocess.content as Reconstructed<T, E>).content,
                parentToPostprocess.links
            ).forEach { result.addChild(it.node, it.distance) }
            return result
        }

        private fun postprocess(
            parentContent: E, links: List<NodeLink<ObservedOrReconstructed<T, E>>>
        ): List<NodeLink<ObservedOrReconstructed<T, E>>> = links
            .flatMap { link ->
                val child = link.node
                when (child.content) {
                    is Reconstructed<*, *> -> {
                        val childContent = child.content as Reconstructed<T, E>
                        val mapped = postprocessDescendants(parentContent, childContent.content)
                        val newDistance = distance(parentContent, mutationsBetween(parentContent, mapped))
                        if (newDistance == 0.0) {
                            //actually parent equals mapped result.
                            //in this case we should remove this node from tree by skipping it and inserting its children directly
                            postprocess(parentContent, child.links)
                        } else {
                            val result = Tree.Node<ObservedOrReconstructed<T, E>>(
                                Reconstructed(
                                    mapped,  //TODO recalculate minDistanceFromObserved
                                    childContent.minDistanceFromObserved,
                                    ++counter
                                )
                            )
                            postprocess(mapped, child.links).forEach {
                                result.addChild(it.node, it.distance)
                            }
                            listOf(NodeLink(result, newDistance))
                        }
                    }
                    else -> listOf(link)
                }
            }
    }

    sealed class ObservedOrReconstructed<T, E> constructor(val id: Int) {
        fun <R1, R2> map(
            mapObserved: (T) -> R1,
            mapReconstructed: (E) -> R2
        ): ObservedOrReconstructed<R1, R2> = when (this) {
            is Observed -> Observed(mapObserved(content), id)
            is Reconstructed -> Reconstructed(mapReconstructed(content), minDistanceFromObserved, id)
        }

        fun <R> convert(mapObserved: (T) -> R, mapReconstructed: (E) -> R): R = when (this) {
            is Observed -> mapObserved(content)
            is Reconstructed -> mapReconstructed(content)
        }
    }

    internal class Observed<T, E>(
        val content: T,
        id: Int
    ) : ObservedOrReconstructed<T, E>(id)

    internal class Reconstructed<T, E>(
        val content: E,
        /**
         * This number represents distance that will be added after removing of all reconstructed nodes from the tree.
         *
         *
         * zero for root
         * zero if had an observed child with distance zero
         * otherwise it is a distance from the nearest parent with minDistanceFromObserved equals zero
         */
        var minDistanceFromObserved: Double,
        id: Int
    ) : ObservedOrReconstructed<T, E>(id)

    companion object {
        val ACTION_COMPARATOR: Comparator<Action<*>> = Comparator
            //TODO check why and revive tests with 0 and 1
            .comparingDouble<Action<*>> { action ->
                action.changeOfDistance + action.distanceFromObserved
            }
            .thenComparing { action -> action.changeOfDistance }
            .thenComparing { action -> if (action is TreeBuilderByAncestors<*, *, *>.Insert) 1 else -1 }
    }
}
